File: iris-crypto/iris-crypto-rsa/src/main/java/city/smartb/iris/crypto/rsa/RSAKeyPairReader.java
package city.smartb.iris.crypto.rsa;

import city.smartb.iris.crypto.rsa.exception.InvalidRsaKeyException;
import city.smartb.iris.crypto.rsa.utils.FileUtils;
import org.bouncycastle.asn1.pkcs.RSAPrivateKey;
import org.bouncycastle.util.io.pem.PemObject;
import org.bouncycastle.util.io.pem.PemReader;

import java.io.IOException;
import java.io.Reader;
import java.security.*;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.RSAPrivateCrtKeySpec;
import java.security.spec.X509EncodedKeySpec;

public class RSAKeyPairReader {

    public static KeyPair loadKeyPair(String filename) throws InvalidRsaKeyException {
        PrivateKey priv = loadPrivateKey(filename);
        PublicKey pub = loadPublicKey(filename);
        return new KeyPair(pub, priv);
    }

    public static PrivateKey loadPrivateKey(String filename) throws InvalidRsaKeyException {
        try {
            PemObject pem = getPemObject(filename);
            RSAPrivateKey key = RSAPrivateKey.getInstance(pem.getContent());
            KeyFactory kf = KeyFactory.getInstance("RSA");
            RSAPrivateCrtKeySpec privSpec = new RSAPrivateCrtKeySpec(key.getModulus(), key.getPublicExponent(), key.getPrivateExponent(), key.getPrime1(), key.getPrime2(), key.getExponent1(), key.getExponent2(), key.getCoefficient());
            return kf.generatePrivate(privSpec);
        } catch (InvalidKeySpecException | NoSuchAlgorithmException | IOException e) {
            throw new InvalidRsaKeyException("Private key can't be loaded", e);
        }
    }

    public static PublicKey loadPublicKey(String filename) throws InvalidRsaKeyException {
        try {
            if (!filename.endsWith(".pub")) {
                filename = filename.concat(".pub");
            }
            PemObject pem = getPemObject(filename);
            byte[] pubKeyBytes = pem.getContent();

            X509EncodedKeySpec pubSpec = new X509EncodedKeySpec(pubKeyBytes);
            KeyFactory kf = KeyFactory.getInstance("RSA");
            return kf.generatePublic(pubSpec);
        } catch (InvalidKeySpecException | NoSuchAlgorithmException | IOException e) {
            throw new InvalidRsaKeyException("Public key can't be loaded", e);
        }
    }


    public static PublicKey fromByteArray(byte[] pub) throws InvalidRsaKeyException {
        try {
            return KeyFactory.getInstance("RSA").generatePublic(new X509EncodedKeySpec(pub));
        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
            throw new InvalidRsaKeyException("Public key can't be loaded", e);
        }
    }

    public static KeyPair generateRSAKey() throws InvalidRsaKeyException {
        try {
            KeyPairGenerator kpg = KeyPairGenerator.getInstance("RSA");
            kpg.initialize(2048);
            return kpg.generateKeyPair();
        } catch (NoSuchAlgorithmException e) {
            throw new InvalidRsaKeyException("RSA key can't be generated", e);
        }
    }

    private static PemObject getPemObject(String filename) throws IOException {
        Reader reader = FileUtils.getReader(filename);
        PemReader rpem = new PemReader(reader);
        return rpem.readPemObject();
    }

}

File: iris-crypto/iris-crypto-rsa-key/src/main/kotlin/city/smartb/iris/crypto/rsa/RSAKeyPairReader.kt
package city.smartb.iris.crypto.rsa

import city.smartb.iris.crypto.rsa.exception.InvalidRsaKeyException
import city.smartb.iris.crypto.rsa.utils.FileUtils
import java.io.IOException
import java.io.Reader
import java.security.KeyFactory
import java.security.KeyPair
import java.security.KeyPairGenerator
import java.security.NoSuchAlgorithmException
import java.security.PrivateKey
import java.security.PublicKey
import java.security.spec.InvalidKeySpecException
import java.security.spec.RSAPrivateCrtKeySpec
import java.security.spec.X509EncodedKeySpec
import org.bouncycastle.asn1.pkcs.RSAPrivateKey
import org.bouncycastle.util.io.pem.PemObject
import org.bouncycastle.util.io.pem.PemReader

object RSAKeyPairReader {

    @Throws(InvalidRsaKeyException::class)
    fun loadKeyPair(filename: String): KeyPair {
        val priv = loadPrivateKey(filename)
        val pub = loadPublicKey(filename)
        return KeyPair(pub, priv)
    }

    @Throws(InvalidRsaKeyException::class)
    fun loadPrivateKey(filename: String): PrivateKey {
        return try {
            val pem = getPemObject(filename)
            val key = RSAPrivateKey.getInstance(pem.content)
            val kf = KeyFactory.getInstance("RSA")
            val privSpec = RSAPrivateCrtKeySpec(
                key.modulus,
                key.publicExponent,
                key.privateExponent,
                key.prime1,
                key.prime2,
                key.exponent1,
                key.exponent2,
                key.coefficient
            )
            kf.generatePrivate(privSpec)
        } catch (e: InvalidKeySpecException) {
            throw InvalidRsaKeyException("Private key can't be loaded", e)
        } catch (e: NoSuchAlgorithmException) {
            throw InvalidRsaKeyException("Private key can't be loaded", e)
        } catch (e: IOException) {
            throw InvalidRsaKeyException("Private key can't be loaded", e)
        }
    }

    @Throws(InvalidRsaKeyException::class)
    fun loadPublicKey(filename: String): PublicKey {
        var filename = filename
        return try {
            if (!filename.endsWith(".pub")) {
                filename = "$filename.pub"
            }
            val pem = getPemObject(filename)
            val pubKeyBytes = pem.content
            val pubSpec = X509EncodedKeySpec(pubKeyBytes)
            val kf = KeyFactory.getInstance("RSA")
            kf.generatePublic(pubSpec)
        } catch (e: InvalidKeySpecException ) {
            throw InvalidRsaKeyException("Public key can't be loaded", e)
        } catch (e: NoSuchAlgorithmException) {
            throw InvalidRsaKeyException("Public key can't be loaded", e)
        } catch (e: IOException) {
            throw InvalidRsaKeyException("Public key can't be loaded", e)
        }
    }

    @Throws(InvalidRsaKeyException::class)
    fun fromByteArray(pub: ByteArray?): PublicKey {
        return try {
            KeyFactory.getInstance("RSA").generatePublic(X509EncodedKeySpec(pub))
        } catch (e: InvalidKeySpecException) {
            throw InvalidRsaKeyException("Public key can't be loaded", e)
        } catch (e: NoSuchAlgorithmException) {
            throw InvalidRsaKeyException("Public key can't be loaded", e)
        }
    }

    @Throws(InvalidRsaKeyException::class)
    fun generateRSAKey(): KeyPair {
        return try {
            val kpg = KeyPairGenerator.getInstance("RSA")
            kpg.initialize(2048)
            kpg.generateKeyPair()
        } catch (e: NoSuchAlgorithmException) {
            throw InvalidRsaKeyException("RSA key can't be generated", e)
        }
    }

    @Throws(IOException::class)
    private fun getPemObject(filename: String): PemObject {
        val reader: Reader = FileUtils.getReader(filename)
        val rpem = PemReader(reader)
        return rpem.readPemObject()
    }
}
